### PR TITLE
feat: make federation info available before joining

### LINKED ISSUE #3514: Separate config fetching from building the client
There should be a struct that just contains the federation config and can be created from the join code, without initializing a full client. This can be used to:

* Preview Federation info
* Create a API client that talks to all the guardian servers
* Give it to a client builder to instantiate a cient

### PR DESCRIPTION
**Pull request by elsirion**
Fixes #3514

### CODE REVIEW

File: fedimint-client/src/lib.rs [hunk 1]
```diff
@@ -1258,21 +1258,77 @@ impl TransactionUpdates {
...
+        Ok(FederationInfo {
+            config,
+            invite_code: None,
+        })
+    }
+
+    /// Returns the federations configuration
+    pub fn config(&self) -> &ClientConfig {
+        &self.config
+    }
```
Comment (elsirion) [inline fedimint-client/src/lib.rs:1294]: Do we actually want to make that available or should there be functions for all the relevant fields? @wbobeirne @justinmoon

File: fedimint-client/src/lib.rs [hunk 2]
```diff
@@ -1258,21 +1258,77 @@ impl TransactionUpdates {
...
+pub struct FederationInfo {
+    config: ClientConfig,
+    // TODO: make non-optional or remove
+    invite_code: Option<InviteCode>,
+}
+
+impl FederationInfo {
+    /// Download federation info using invitation code
+    pub async fn from_invite_code(invite: InviteCode) -> anyhow::Result<FederationInfo> {
+        let config = try_download_config(invite.clone(), 10).await?;
```
Comment (elsirion) [inline fedimint-client/src/lib.rs:1273]: I just kept the old semantics, but maybe we should make the retry-behavior configurable?

File: fedimint-client/src/lib.rs [hunk 3]
```diff
@@ -1258,21 +1258,77 @@ impl TransactionUpdates {
...
+            invite_code: None,
+        })
+    }
+
+    /// Returns the federations configuration
+    pub fn config(&self) -> &ClientConfig {
+        &self.config
+    }
+
+    /// If the federation info was created from
```
Comment (okjodom) [inline fedimint-client/src/lib.rs:1296]: nit: dangling comment?

File: fedimint-client/src/lib.rs (hunk 2 above)
Comment (okjodom) [reply fedimint-client/src/lib.rs:1273]: I previously thought the client builder would take config for retries. Makes sense for me

Comment (justinmoon) [reply fedimint-client/src/lib.rs:1273]: https://github.com/fedimint/fedimint/issues/3519

File: fedimint-client/src/lib.rs (hunk 1 above)
Comment (justinmoon) [reply fedimint-client/src/lib.rs:1294]: I think just spitting up the whole config is best for now

**Review by justinmoon — APPROVED** [review]
LGTM

**Review by okjodom — DISMISSED** [review]

File: fedimint-client/src/lib.rs (hunk 3 above)
Comment (elsirion) [reply fedimint-client/src/lib.rs]: Completed my comment

**Comment by elsirion** [comment]
Oh, I didn't see that you already queued it :laughing:

**Pull request merged**